From 69cbb10d76aacf304cff74561c63dc165fb3f31c Mon Sep 17 00:00:00 2001
From: Jianwei Fan <jianwei.fan@rock-chips.com>
Date: Tue, 2 Jan 2024 11:23:14 +0000
Subject: [PATCH] media: i2c: rk628: add register debugfs node

This patch adds support for register debugfs nodes to allow access to
the registers without modifying the regmap subsystem configuration.

for example:

read rk628 csi registers:
        cat /d/rk628/3-0050/registers/csi

write rk628 csi register:
        echo addr val > /d/rk628/3-0050/registers/csi

Change-Id: I1bb4b4acff939718eca43fc24c25cf673726cb4c
Signed-off-by: Jianwei Fan <jianwei.fan@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628.c          | 119 +++++++++++++++++++++++
 drivers/media/i2c/rk628/rk628.h          |   2 +
 drivers/media/i2c/rk628/rk628_csi_v4l2.c |   3 +
 3 files changed, 124 insertions(+)

diff --git a/drivers/media/i2c/rk628/rk628.c b/drivers/media/i2c/rk628/rk628.c
index 79d712523992cf..9681becba5f9a0 100644
--- a/drivers/media/i2c/rk628/rk628.c
+++ b/drivers/media/i2c/rk628/rk628.c
@@ -430,6 +430,125 @@ static const struct regmap_config rk628_regmap_config[RK628_DEV_MAX] = {
 	},
 };
 
+static int rk628_reg_show(struct seq_file *s, void *v)
+{
+	const struct regmap_config *reg;
+	struct rk628 *rk628 = s->private;
+	unsigned int i, j;
+	u32 val;
+
+	seq_printf(s, "rk628_%s:\n", file_dentry(s->file)->d_iname);
+
+	for (i = 0; i < ARRAY_SIZE(rk628_regmap_config); i++) {
+		reg = &rk628_regmap_config[i];
+		if (!reg->name)
+			continue;
+		if (!strcmp(reg->name, file_dentry(s->file)->d_iname))
+			break;
+	}
+
+	if (i == ARRAY_SIZE(rk628_regmap_config))
+		return -ENODEV;
+
+	/* grf */
+	if (!reg->rd_table) {
+		for (i = 0; i <= reg->max_register; i += 4) {
+			rk628_i2c_read(rk628, i, &val);
+			if (i % 16 == 0)
+				seq_printf(s, "\n0x%04x:", i);
+			seq_printf(s, " %08x", val);
+		}
+	} else {
+		const struct regmap_range *range_list = reg->rd_table->yes_ranges;
+		const struct regmap_range *range;
+		int range_list_len = reg->rd_table->n_yes_ranges;
+
+		for (i = 0; i < range_list_len; i++) {
+			range = &range_list[i];
+			for (j = range->range_min; j <= range->range_max; j += 4) {
+				rk628_i2c_read(rk628, j, &val);
+				if (j % 16 == 0 || j == range->range_min)
+					seq_printf(s, "\n0x%04x:", j);
+				seq_printf(s, " %08x", val);
+			}
+		}
+
+		seq_puts(s, "\n");
+	}
+
+	return 0;
+}
+
+static ssize_t rk628_reg_write(struct file *file, const char __user *buf,
+			       size_t count, loff_t *ppos)
+{
+	struct rk628 *rk628 = file->f_path.dentry->d_inode->i_private;
+	u32 addr;
+	u32 val;
+	char kbuf[25];
+	int ret;
+
+	if (count >= sizeof(kbuf))
+		return -ENOSPC;
+
+	if (copy_from_user(kbuf, buf, count))
+		return -EFAULT;
+
+	kbuf[count] = '\0';
+
+	ret = sscanf(kbuf, "%x%x", &addr, &val);
+	if (ret != 2)
+		return -EINVAL;
+
+	rk628_i2c_write(rk628, addr, val);
+
+	return count;
+}
+
+static int rk628_reg_open(struct inode *inode, struct file *file)
+{
+	struct rk628 *rk628 = inode->i_private;
+
+	return single_open(file, rk628_reg_show, rk628);
+}
+
+static const struct file_operations rk628_reg_fops = {
+	.owner          = THIS_MODULE,
+	.open           = rk628_reg_open,
+	.read           = seq_read,
+	.write          = rk628_reg_write,
+	.llseek         = seq_lseek,
+	.release        = single_release,
+};
+
+static void rk628_debugfs_register_create(struct rk628 *rk628)
+{
+	const struct regmap_config *reg;
+	struct dentry *dir;
+	int i;
+
+	dir = debugfs_create_dir("registers", rk628->debug_dir);
+	if (IS_ERR(dir))
+		return;
+
+	for (i = 0; i < ARRAY_SIZE(rk628_regmap_config); i++) {
+		reg = &rk628_regmap_config[i];
+		if (!reg->name)
+			continue;
+		debugfs_create_file(reg->name, 0600, dir, rk628, &rk628_reg_fops);
+	}
+}
+
+void rk628_debugfs_create(struct rk628 *rk628)
+{
+	rk628->debug_dir = debugfs_create_dir(dev_name(rk628->dev), debugfs_lookup("rk628", NULL));
+	if (IS_ERR(rk628->debug_dir))
+		return;
+
+	rk628_debugfs_register_create(rk628);
+}
+EXPORT_SYMBOL(rk628_debugfs_create);
+
 struct rk628 *rk628_i2c_register(struct i2c_client *client)
 {
 	struct rk628 *rk628;
diff --git a/drivers/media/i2c/rk628/rk628.h b/drivers/media/i2c/rk628/rk628.h
index e37f742d6f7170..98652d5c9eb32b 100644
--- a/drivers/media/i2c/rk628/rk628.h
+++ b/drivers/media/i2c/rk628/rk628.h
@@ -302,6 +302,7 @@ struct rk628 {
 	bool dual_mipi;
 	struct mipi_timing mipi_timing[2];
 	int tx_mode;
+	struct dentry *debug_dir;
 };
 
 #ifndef RK628_REGMAP
@@ -349,6 +350,7 @@ void rk628_post_process_en(struct rk628 *rk628,
 			   struct videomode *dst,
 			   u64 *dst_pclk);
 void rk628_version_parse(struct rk628 *rk628);
+void rk628_debugfs_create(struct rk628 *rk628);
 
 #include "rk628_grf.h"
 #include "rk628_gpio.h"
diff --git a/drivers/media/i2c/rk628/rk628_csi_v4l2.c b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
index 5e5450514018f5..e2dd8665c93596 100644
--- a/drivers/media/i2c/rk628/rk628_csi_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
@@ -2738,6 +2738,8 @@ static int rk628_csi_probe(struct i2c_client *client,
 		goto err_work_queues;
 	}
 	csi->rk628->dual_mipi = false;
+	rk628_debugfs_create(csi->rk628);
+
 	v4l2_info(sd, "%s found @ 0x%x (%s)\n", client->name,
 		  client->addr << 1, client->adapter->name);
 
@@ -2761,6 +2763,7 @@ static int rk628_csi_remove(struct i2c_client *client)
 {
 	struct rk628_csi *csi = i2c_get_clientdata(client);
 
+	debugfs_remove_recursive(csi->rk628->debug_dir);
 	if (!csi->hdmirx_irq) {
 		del_timer_sync(&csi->timer);
 		flush_work(&csi->work_i2c_poll);
-- 
2.17.1

